### PR TITLE
Improved lambdify printing for numpy

### PR DESCRIPTION
**Pull request by richardotis**
This PR adds a `NumPyPrinter` class. It adds the following features to `lambdify()` when using the `numpy` module:
- Prints `Piecewise` objects as `numpy.select()` calls
- Matrix multiplication as chained `ndarray.dot()` calls (thanks @jcrist)
- Proper logical operator handling (`logical_and`, `logical_or`, etc.)

After discussion with @asmeurer and @certik, this PR also ends the deprecation cycle started by sympy/sympy#7853 so that `ImmutableMatrix` now prints as `numpy.array` by default. It is possible to get the old behavior by passing `[{'ImmutableMatrix': numpy.matrix}, 'numpy']` to the `modules` kwarg. Tests for this and everything listed above are included.

Feedback/review is very desirable. It was a pleasure sprinting with many members of the SymPy team, and I look forward to ongoing collaboration.

### CODE REVIEW

File: sympy/printing/lambdarepr.py [hunk 1]
```diff
@@ -69,6 +69,59 @@ def _print_BooleanTrue(self, expr):
...
+        "Piecewise function printer"
+        # Print tuples here instead of lists because numba may add support
+        #     for select in nopython mode; see numba#1313 on github.
+        exprs = '({0},)'.format(','.join(self._print(arg.expr) for arg in expr.args))
+        conds = '({0},)'.format(','.join(self._print(arg.cond) for arg in expr.args))
+        # If (default_value, True) is a (expr, cond) tuple in a Piecewise object
+        #     it will behave the same as passing the 'default' kwarg to select()
+        #     *as long as* it is the last element in expr.args.
+        # If this is not the case, it may be triggered prematurely.
+        return 'select({0}, {1})'.format(conds, exprs)
```
Comment (asmeurer) [inline sympy/printing/lambdarepr.py]: The SymPy default for Piecewise is `nan`, whereas you told me that select defaults to 0. Should we make this match SymPy?

**Comment by asmeurer** [comment]
I only have the one comment.

**Comment by richardotis** [comment]
I've incorporated the suggested changes from @asmeurer.

Any comments @jcrist?

(Sidenote: Is the SymPy default condition for Piecewise to return a `nan` object or to raise `ValueError`? The documentation isn't clear on that. When I call `subs()` on a `Piecewise` object where all conditions are False, I get back an empty `Piecewise`.)

**Comment by certik** [comment]
@richardotis thanks a lot for working on this. These are very useful additions to SymPy.

This looks good to me. Tests pass. @jcrist do you have any comments to this?

If not, we can merge this.

**Comment by jcrist** [comment]
Looks good to me. Thanks!

**Pull request merged**